### PR TITLE
Build dependency libraries instead of bundling pre-built binaries.

### PR DESCRIPTION
**Pull request by binary1248**
Title.

In order to:
- Improve toolchain compatibility (e.g. building with Emscripten)
- Reduce maintenance effort of having to build and ship binaries
- Reduce the size of the source archive
- Allow more interesting dependencies (e.g. HarfBuzz) to be easily integrated in the future
- Provide a better debugging experience by providing debug libraries and symbols of dependencies
- Allow user compiler/linker flags to propagate down to the dependencies as well

This change gets rid of the packaged pre-built dependencies and builds them alongside the SFML modules.

It would be nice to get this into SFML 3 so that we have a clean environment to start adding some bigger features in SFML 4.

Currently the focus was on getting this to build successfully on CI, however more widespread testing will be necessary especially on more exotic platforms that we don't currently cover in our CI. The goal is to support as many platforms as is feasible as long as they are still considered somewhat up-to-date.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -106,39 +106,9 @@ if(SFML_BUILD_GRAPHICS AND NOT SFML_BUILD_WINDOW)
     set(SFML_BUILD_WINDOW TRUE)
 endif()
 
-# allow not using bundled dependencies with a switch
-# (except for stb_image)
-# yes this is horrible, but GLOB_RECURSE sucks
 sfml_set_option(SFML_USE_SYSTEM_DEPS FALSE BOOL "TRUE to use system dependencies, FALSE to use the bundled ones.")
```
Comment (MarioLiebisch) [inline CMakeLists.txt:109]: Maybe with this change it would be better to default this to `TRUE` (maybe only on platforms other than Windows)? Or it could essentially be a tri-state between "Only use system dependencies", "Build all dependencies" or "Build only missing dependencies"?

Comment (binary1248) [reply CMakeLists.txt:109]: I think we have to look at it from a user experience perspective.

Experienced users are going to know this option exists and know whether they want to build using system dependencies or not and what that means in terms of requirements e.g. installing development libraries, having to install packages on target systems as well etc.

Beginner users who just wanted to get access to the master version of SFML might not know what using system dependencies means and if it is even viable on the system they are developing on/for.

I think leaving the default at a value that is going to work better for beginner users is the better choice.

Comment (texus) [reply CMakeLists.txt:109]: Didn't it used system libraries on Linux before this PR? (e.g. [src/SFML/Graphics/CMakeLists.txt#L101-L110](https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/CMakeLists.txt#L101-L110) didn't set the include directory on Linux)

I haven't taken a proper look at this PR yet, but I imagine that you would then also need to install all the dependencies such as freetype to /usr/local/ in order for the application to use SFML? Couldn't it cause issues for other projects that there is suddenly a freetype include and lib available in /usr/local that overrules the system one in /usr? (On Windows that would be less of a problem because you install to some custom folder that doesn't affect the rest of the system).

If you already considered the consequences of switching to bundled libraries on linux and you find that there shouldn't be any problems with it with either static or dynamic linking then I guess it's probably fine to use bundled libs by default.

Comment (binary1248) [reply CMakeLists.txt:109]: I guess SFML did always use system dependencies on Linux. Now that I see it I don't think it is very intuitive to provide `SFML_USE_SYSTEM_DEPS` and for it not to have an effect when building on Linux. Since we can't make any assumptions on systems we haven't yet encountered, the default for `SFML_USE_SYSTEM_DEPS` should still be `FALSE` but it would make sense to set the default to `TRUE` on Linux to mimic the old behaviour.

>  haven't taken a proper look at this PR yet, but I imagine that you would then also need to install all the dependencies such as freetype to /usr/local/ in order for the application to use SFML? Couldn't it cause issues for other projects that there is suddenly a freetype include and lib available in /usr/local that overrules the system one in /usr? (On Windows that would be less of a problem because you install to some custom folder that doesn't affect the rest of the system).

/usr/local should never be touched by the system package manager. It is the place that "local system configuration" i.e. self-built libraries and configuration files should be placed into. It is up to the user to maintain it. I've already built many things that ended up being installed to /usr/local by default. Because the search paths are set to prioritize /usr/local it will always be preferred over anything else if the user doesn't explicitly specify anything.

Comment (eXpl0it3r) [reply CMakeLists.txt]: I guess since we still require some OS dependencies (X11, udev, etc.) on Linux, it should be fine to also recommend installing the system version for our dependencies.

With `brew` being the defacto package manager on macOS, it's not official and doesn't get shipped by default, so we can't make the same assumption.  
Windows is actually further ahead with `winget` being officially supported by Microsoft and coming pre-installed (on Windows 11?), but then again Windows doesn't really have a (useful) way to install development libraries and also doesn't seem to have pure libraries.

File: src/SFML/Audio/CMakeLists.txt [hunk 2]
```diff
@@ -67,16 +67,93 @@ if(SFML_OS_IOS)
...
+        set(WITH_FORTIFY_SOURCE OFF)
+        set(WITH_STACK_PROTECTOR OFF)
+        set(WITH_AVX OFF) # LLVM/Clang on Windows has issues with AVX2
+
+        FetchContent_Declare(ogg
+            GIT_REPOSITORY https://github.com/xiph/ogg.git
+            GIT_TAG v1.3.5
+            GIT_SHALLOW ON
+            # patch CMAKE_DEBUG_POSTFIX into the Ogg CMake configuration
+            PATCH_COMMAND ${CMAKE_COMMAND} -DOGG_DIR=${FETCHCONTENT_BASE_DIR}/ogg-src -P ${CMAKE_SOURCE_DIR}/tools/ogg/PatchOgg.cmake)
```
Comment (vittorioromeo) [inline src/SFML/Audio/CMakeLists.txt:103]:
```suggestion
            PATCH_COMMAND ${CMAKE_COMMAND} -DOGG_DIR=${FETCHCONTENT_BASE_DIR}/ogg-src -P ${CMAKE_CURRENT_LIST_DIR}/../../../tools/ogg/PatchOgg.cmake)
```

I've had to change these lines here, otherwise projects adding SFML itself via `FetchContent` would not find the `.cmake` file. Not sure what the idiomatic solution is, but this worked...

**Review by vittorioromeo — CHANGES_REQUESTED** [review]

**Comment by vittorioromeo** [comment]
Another issue I am getting is that, when building `imgui-sfml` using

```
cmake .. -DIMGUI_SFML_FIND_SFML=1 -DSFML_ROOT="C:/OHWorkspace/SFML/build/install" -DIMGUI_DIR="C:/OHWorkspace/imgui" -DSFML_DIR="C:/OHWorkspace/SFML/build/install"
```

I get the following error

```
Call Stack (most recent call first):
  CMakeLists.txt:28 (find_package)


CMake Error at CMakeLists.txt:28 (find_package):
  Found package configuration file:

    C:/OHWorkspace/SFML/build/install/lib/cmake/SFML/SFMLConfig.cmake

  but it set SFML_FOUND to FALSE so package "SFML" is considered to be NOT
  FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  cpptrace::cpptrace Vorbis::vorbis Vorbis::vorbisfile Vorbis::vorbisenc
  FLAC::FLAC
```

Not sure what the right way of solving this one is.

File: src/SFML/Audio/CMakeLists.txt (hunk 2 above)
Comment (ChrisThrasher) [reply src/SFML/Audio/CMakeLists.txt:103]: `PROJECT_SOURCE_DIR` is the idiomatic solution.

**Comment by ChrisThrasher** [comment]
> I get the following error

Likewise. Not sure what the fix is either.

**Comment by ChrisThrasher** [comment]
Oddly enough, I can fix that ImGui-SFML configuration error by adding the `Audio` component to the `find_package(SFML` call. Perhaps something about package component handling got broken.

File: cmake/SFMLConfig.cmake.in [hunk 3]
```diff
@@ -121,8 +121,8 @@ set(targets_config_file "${CMAKE_CURRENT_LIST_DIR}/SFML${config_name}Targets.cma
 if(EXISTS "${targets_config_file}")
     # Set SFML_FOUND to ON by default, may be overwritten by one of the includes below
     set(SFML_FOUND ON)
-    include("${targets_config_file}")
     include("${CMAKE_CURRENT_LIST_DIR}/SFMLConfigDependencies.cmake")
+    include("${targets_config_file}")
```
Comment (ChrisThrasher) [inline cmake/SFMLConfig.cmake.in:125]: What was the purpose of reordering these two lines?

Comment (binary1248) [reply cmake/SFMLConfig.cmake.in:125]: Wouldn't build otherwise. `SFMLConfigDependencies.cmake` depends on stuff that is introduced in `targets_config_file`.

**Comment by ChrisThrasher** [comment]
Have you looked into CPM.cmake and whether it would offer a simpler implementation? For example, CPM.cmake will let you specify options for fetched projects that you do stuff like build a dependency as a static library while not affecting the value of `BUILD_SHARED_LIBS` in the consuming project. This removes the need to cache the value of that variable then careful restore it later.

**Comment by binary1248** [comment]
The initial version of this change used CPM.cmake and as much as I would have liked to keep using it it very quickly falls apart when trying to add dependencies that can't be bothered to adhere to standard CMake practices. The fact that we have to patch many dependencies should already be a red flag that CPM.cmake isn't going to have a good time with them.

**Comment by eXpl0it3r** [comment]
@binary1248 if have a bit of time, could you look at what's causing the error that is mentioned by Vittorio?

I really like to include this for the next RC, so that we have more than the SFML Team to try out this change.

In my opinion this is on the critical path. For one we've decided to move in this direction many years ago, but more importantly this will unlock a lot of new possibilities, things that have so far held back development for certain features like HarfBuzz or Opus support.

I don't think this should be held back for SFML 3.1 or similar, as it's quite a fundamental change in the dependency handling and will cause more issues, if users have to migrate between minor versions.

**Comment by binary1248** [comment]
If someone could add the appropriate tests for this scenario to CI it would be easier to ensure this change or any future changes don't break this functionality. It would also make it easier for me to fix the outstanding issue/issues without having to make sure my local test environment matches what is expected.

**Comment by ChrisThrasher** [comment]
> If someone could add the appropriate tests for this scenario to CI it would be easier to ensure this change or any future changes don't break this functionality. It would also make it easier for me to fix the outstanding issue/issues without having to make sure my local test environment matches what is expected.

I added those tests two years ago. 
https://github.com/SFML/SFML/pull/2206

> Oddly enough, I can fix that ImGui-SFML configuration error by adding the Audio component to the find_package(SFML call. Perhaps something about package component handling got broken.

The problem is that this error only occurs if you omit the `Audio` component when calling `find_package`. Our install interface tests requests all components.

https://github.com/SFML/SFML/blob/e21ba4dc950bbe49a676d1dd5c8c1b3f604a438f/test/install/CMakeLists.txt#L16

**Comment by ChrisThrasher** [comment]
> I don't think this should be held back for SFML 3.1 or similar, as it's quite a fundamental change in the dependency handling and will cause more issues, if users have to migrate between minor versions.

Is it? Notice how all our CI build jobs still work without modification. What about users' workflows will change if we merge this?

**Comment by eXpl0it3r** [comment]
I narrowed it down to the difference between using `SFML_DIR` and `SFML_ROOT`, the former we use in the install tests, thus it doesn't fail in CI.

Simple repo:

1. Build & install SFML to some directory:
    - `cmake -S . -B build -DCMAKE_INSTALL_PREFIX=C:/Temp/SFML`
    - `cmake --build build --target install`
2. Create the two files from below
3. Run CMake for the project: `cmake -S . -B build -DSFML_DIR=C:/Temp/SFML/lib/cmake/SFML`
4. Fail

It works if you run instead `cmake -S . -B build -DSFML_ROOT=C:/Temp/SFML`

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.28)
project(CMakeSFMLProject LANGUAGES CXX)

set(SFML_STATIC_LIBRARIES ON)
find_package(SFML REQUIRED COMPONENTS Graphics Audio)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE SFML::Graphics SFML::Audio)
```

main.cpp
```cpp
#include <SFML/Graphics.hpp>

int main()
{
    auto window = sf::RenderWindow({1920u, 1080u}, "CMake SFML Project");
    window.setFramerateLimit(144);

    while (window.isOpen())
    {
        for (auto event = sf::Event(); window.pollEvent(event);)
        {
            if (event.type == sf::Event::Closed)
            {
                window.close();
            }
        }

        window.clear();
        window.display();
    }
}
```

**Comment by eXpl0it3r** [comment]
I don't know if there is a solution to allow `SFML_DIR`.

Personally, I'm also fine to merge this as is and discourage usage of `SFML_DIR` in favor of `SFML_ROOT`.

**Comment by eXpl0it3r** [comment]
Also, I asked on CMake's Discourse: https://discourse.cmake.org/t/problem-with-fetchcontent-dependencies-when-using-xyz-dir-xyz-root-works/12987

File: cmake/Macros.cmake [hunk 4]
```diff
@@ -218,13 +218,42 @@ macro(sfml_add_library module)
...
 
+    # because the frameworks directory hierarchy has to be set up before any target files
+    # are installed we can't call install(EXPORT ...Targets) here
+    # this is because frameworks are only set up after all modules directories have already been added
+    # sfml_export_targets() is called after the frameworks are set up so we will have to
+    # save all modules to a global property and read it out to call install(EXPORT ...Targets)
+    # for each module in sfml_export_targets(), see below
+    get_property(SFML_ADD_LIBRARY_MODULES GLOBAL PROPERTY SFML_ADD_LIBRARY_MODULES_PROPERTY)
+    list(APPEND SFML_ADD_LIBRARY_MODULES ${module})
+    set_property(GLOBAL PROPERTY SFML_ADD_LIBRARY_MODULES_PROPERTY "${SFML_ADD_LIBRARY_MODULES}")
```
Comment (eXpl0it3r) [inline cmake/Macros.cmake:242]: Why are we using global properties here?

File: cmake/Modules/FindFreetype.cmake [hunk 5]
```diff
@@ -148,12 +148,16 @@ find_package_handle_standard_args(Freetype DEFAULT_MSG FREETYPE_LIBRARY FREETYPE
...
+  add_library(freetype IMPORTED UNKNOWN)
+  set_target_properties(freetype PROPERTIES INTERFACE_INCLUDE_DIRECTORIES "${FREETYPE_INCLUDE_DIRS}")
   if(FREETYPE_LIBRARY MATCHES "/([^/]+)\\.framework$")
-    set_target_properties(Freetype::Freetype PROPERTIES IMPORTED_LOCATION ${FREETYPE_LIBRARY}/${CMAKE_MATCH_1})
+    set_target_properties(freetype PROPERTIES IMPORTED_LOCATION ${FREETYPE_LIBRARY}/${CMAKE_MATCH_1})
   else()
-    set_target_properties(Freetype::Freetype PROPERTIES IMPORTED_LOCATION ${FREETYPE_LIBRARY})
+    set_target_properties(freetype PROPERTIES IMPORTED_LOCATION ${FREETYPE_LIBRARY})
   endif()
 endif()
```
Comment (eXpl0it3r) [inline cmake/Modules/FindFreetype.cmake:159]: Why is this change needed here?

File: src/SFML/Graphics/CMakeLists.txt [hunk 6]
```diff
@@ -97,26 +98,57 @@ target_include_directories(sfml-graphics SYSTEM PRIVATE "${PROJECT_SOURCE_DIR}/e
...
+        set(FT_DISABLE_BROTLI ON)
+        set(FT_ENABLE_ERROR_STRINGS ON)
+
+        FetchContent_Declare(Freetype
+            GIT_REPOSITORY https://github.com/freetype/freetype.git
+            GIT_TAG VER-2-13-2
+            GIT_SHALLOW ON)
+        FetchContent_MakeAvailable(Freetype)
+
+        set_target_properties(freetype PROPERTIES FOLDER "Dependencies")
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/CMakeLists.txt:135]: Should this be capitalized?
```suggestion
        set_target_properties(Freetype PROPERTIES FOLDER "Dependencies")
```

File: src/SFML/Graphics/CMakeLists.txt [hunk 7]
```diff
@@ -97,26 +98,57 @@ target_include_directories(sfml-graphics SYSTEM PRIVATE "${PROJECT_SOURCE_DIR}/e
...
+        # as static libraries we need to build them with -fPIC
+        if(SFML_BUILD_SHARED_LIBS)
+            set_target_properties(freetype PROPERTIES POSITION_INDEPENDENT_CODE ON)
+        endif()
+
+        # disable building dependencies as part of a unity build, they don't support it
+        set_target_properties(freetype PROPERTIES UNITY_BUILD OFF)
+
+        sfml_set_stdlib(freetype)
+        add_library(Freetype::Freetype ALIAS freetype)
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/CMakeLists.txt:147]:
```suggestion
        if(SFML_BUILD_SHARED_LIBS)
            set_target_properties(Freetype PROPERTIES POSITION_INDEPENDENT_CODE ON)
        endif()

        # disable building dependencies as part of a unity build, they don't support it
        set_target_properties(Freetype PROPERTIES UNITY_BUILD OFF)

        sfml_set_stdlib(Freetype)
        add_library(Freetype::Freetype ALIAS Freetype)
```

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Generally looks good. Some small comments and I still need to retest it.

File: cmake/Modules/FindFreetype.cmake (hunk 5 above)
Comment (binary1248) [reply cmake/Modules/FindFreetype.cmake:159]: To match what is exported by the FreeType CMake configuration.

File: src/SFML/Graphics/CMakeLists.txt (hunk 6 above)
Comment (binary1248) [reply src/SFML/Graphics/CMakeLists.txt:135]: The target name used by the FreeType CMake configuration is lowercase `freetype`.

Comment (eXpl0it3r) [reply src/SFML/Graphics/CMakeLists.txt:135]: Maybe, I'm misunderstanding something, but shouldn't it match the casing of the FetchContent call?

**Review by eXpl0it3r — APPROVED** [review]

**Comment by eXpl0it3r** [comment]
Tested it successfully locally. It's awesome to know that we no longer depend on pre-built binaries. The build time also doesn't seem to drastically increase for me.

We may want to consider updating some of the dependencies after this PR.

**Comment by eXpl0it3r** [comment]
Remember when Bromeon opened an internal proposal for this in September 2015? 😄 

Let's see what we broke with this change, but I'm excited to see all the possibilities ahead!

**Comment by ChrisThrasher** [comment]
I want to mention a few observations as I update some projects to the latest state of `master`.

1. `project(my_app CXX)` no longer works. SFML requires a C compiler to build and for whatever reason that requirement isn't automatically inherited by projects that consume SFML via `FetchContent`. This generates a rather obtuse warning that I think many users will struggle to understand. Users must use `project(my_app C CXX)` or more simply `project(my_app)` to deal with this. I'm not aware of any action we can take to in the library itself fix this.
    ```
    CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
    Missing variable is:
    CMAKE_C_COMPILE_OBJECT
    CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
    Missing variable is:
    CMAKE_C_COMPILE_OBJECT
    CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
    Missing variable is:
    CMAKE_C_COMPILE_OBJECT
    CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
    Missing variable is:
    CMAKE_C_COMPILE_OBJECT
    CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
    Missing variable is:
    CMAKE_C_COMPILE_OBJECT
    -- Generating done (0.1s)
    CMake Generate step failed.  Build files cannot be regenerated correctly.
    ```

2. Building SFML use to be _very_ clean to nearly any warning you may apply. It's bad practice to build 3rd party code with warnings enabled but sometimes users do it accidentally. Now that we build lots of 3rd party code, enabling even basic warnings causes a lot of console output about SFML's dependencies. Again, no real action we can take but another thing users may notice in the transition.

3. I get a warning in 3rd party code and am not sure how to silence it. This is not a technical problem, but it looks a bit sloppy that building the library produces a warning. Maybe building our C deps with a newer C standard will fix this.
    ```
    /Users/thrasher/Projects/mandelbrot/build/_deps/flac-src/microbench/util.c:79:11: warning: mixing declarations and code is incompatible with standards before C99 [-Wdeclaration-after-statement]
       79 |         uint64_t duration = *end - *start;
          |                  ^
    1 warning generated.
    ```
